src/sortedarray.h

```diff
@@ -0,0 +1,210 @@
+for any purpose with or without fee is hereby granted, provided
+that the above copyright notice and this permission notice appear
+in all copies.
+
+THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
+WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
+AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
+CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
+LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
+NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+ */
+
+/**
+ * @file sortedarray.h
+ *
+ * @brief Automatically sorted and resizing array
+ *
+ * An SortedArray is an automatically resizing sorted array. Most operations
+ * run O(n) worst case running time. Some operations run in O(log n).
+ *
+ * To retrieve a value use the sortedarray structure by accessing the data
+ * field.
+ *
+ * To create a SortedArray, use @ref sortedarray_new
+ * To destroy a SortedArray, use @ref sortedarray_free
+ *
+ * To add a value to a SortedArray, use @ref sortedarray_prepend, 
+ * @ref sortedarray_append, or @ref sortedarray_insert.
+ *
+ * To remove a value from a SortedArray, use @ref sortedarray_remove
+ * or @ref sortedarray_remove_range.
+ */
+
+#ifndef ALGORITHM_SORTEDARRAY_H
+#define ALGORITHM_SORTEDARRAY_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * A value to store in @ref SortedArray.
+ */
+typedef void *SortedArrayValue;
+
+/**
+ * A SortedArray structure. Use @ref sortedarray_new to create one.
+ *
+ * The SortedArray is an automatically resizing array which stores its 
+ * elements in sorted order. Userdefined functions determine the sorting order.
+ * All operations on a SortedArray maintain the sorted property. Most 
+ * operations are done in O(n) time, but searching can be done in O(log n)
+ * worst case.
+ *
+ * @see sortedarray_new
+ */
+typedef struct _SortedArray SortedArray;
+
+/**
+ * Compare two values in a SortedArray to determine if they are equal.
+ *
+ * @param value1	The first value to compare.
+ * @param value2	The second value to compare.
+ * @return		Non-zero if value1 equals value2, zero if they do not
+ *			equal.
+ *
+ */
+typedef int (*SortedArrayEqualFunc)(SortedArrayValue value1,
+                                    SortedArrayValue value2);
+
+/**
+ * Compare two values in a SortedArray to determine their order.
+ *
+ * @param value1	The first value to compare.
+ * @param value2	The second value to compare.
+ * @return		Less than zero if value1 is compared smaller than 
+ * 			value2, zero if they compare equal, or greater than
+ * 			zero if value1 compares greate than value2.
+ */
+typedef int (*SortedArrayCompareFunc)(SortedArrayValue value1,
+                                      SortedArrayValue value2);
+
+/**
+ * Definition of a @ref SortedArray
+ */
+struct _SortedArray {
```

Comment:
In general I prefer to make the structs opaque, ie. declare them inside the .c file and don't expose them publicly. Can you do this please?

You'll need to turn SORTEDARRAY_GET into a proper accessor function, but that's okay - the result should be more readable anyway.